### PR TITLE
Fixes #1378 Update Talawa to give a meaningful message if the URL isn't available.

### LINKED ISSUE #1370: App Stuck at loading , if the url is invalid
**Describe the bug**
Currently after providing , the organisation url , it got stuck at that screen if server is not running.
What I assuming, it just checking the format of url [https://example.com]. 
The app should be responsible to show error and cause to user, and go one step back to refactor the issue.

**To Reproduce**

1. Open app
2. Provide any url [Valid format]
3. App will got stuck at `Select Organization Section`

**Expected behavior**
The app should be responsible to show error and cause to user, and go one step back to refactor the issue.

**Additional details**
- It's better to keep an endpoint in `talawa-api` to check, whether it's an talawa service or not. 
- Also the organization url need to be in this format http://<url>/graphql
- The app should be compaitable , if the user provide http://<url> . It should retry the /graphql in case user has not provided

### LINKED ISSUE #1378: Update Talawa to give a meaningful message if the URL isn't available.
**Is your feature request related to a problem? Please describe.**
When I started the app and followed the setup documentation I ran into this problem, Only got to know about it after I checked out community slack

**Describe the solution you'd like**
Give proper error messages when URl isn't available

**Describe alternatives you've considered**
Give error message

### PR DESCRIPTION
**Pull request by Ayush0Chaudhary**
**What kind of change does this PR introduce?**

**Issue Number:**

Fixes #1378
and kind of fixes #1370 

**Did you add tests for your changes?**
Yes

**Snapshots/Videos:**

https://user-images.githubusercontent.com/95746190/210115449-94795eab-e365-443b-9b34-9596f06f22d8.mp4

**If relevant, did you update the documentation?**
Didn't update the docs.

**Summary**
In this PR if this https://github.com/PalisadoesFoundation/talawa/blob/46cf4efd53cbdbdba3e42633875c667eff4a94cf/lib/widgets/organization_list.dart#L34
`If statement` is true it gives a snackbar after 2sec It will prompt a snackbar. I have to add 2 seconds delay because we cannot prompt a snackbar while the widgets are building. Fix#1368

**Does this PR introduce a breaking change?**
NO

**Other information**
None

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa/blob/master/CONTRIBUTING.md)?**
Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
Please fix the failing tests

File: lib/widgets/organization_list.dart [hunk 1]
```diff
@@ -32,17 +34,10 @@ class OrganizationList extends StatelessWidget {
...
-              showSnackBar: false,
+              showSnackBar: true,
             );
-            if (isException != null) {
-              if (isException) {
-                refetch!();
-              } else {
-                refetch!();
-              }
-            }
```
Comment (SiddheshKukade) [inline lib/widgets/organization_list.dart:45]: Why are these lines removed?

**Review by SiddheshKukade — CHANGES_REQUESTED** [review]

**Comment by SiddheshKukade** [comment]
@Ayush0Chaudhary actually, I can't understand the reason for adding 2-sec delay to the snackBar. We already have a `Verify` button to verify the server URL.

File: lib/widgets/organization_list.dart (hunk 1 above)
Comment (Ayush0Chaudhary) [reply lib/widgets/organization_list.dart:45]: They would have prompted the snackbar again and again.

**Comment by Ayush0Chaudhary** [comment]
It only verify whether the url is working with http.get(uri) but https//:www.youtube.com will also clear this check

But in this we check whether the data received is appropriate for us.

**Comment by Ayush0Chaudhary** [comment]
@SiddheshKukade I added these lines back
https://github.com/PalisadoesFoundation/talawa/pull/1380#discussion_r1059577152

Now the **snackbar will be shown only once** if the url added **doesn't provide the correct data**. Which otherwise would've just shown the loading icon

My goal here was to tell user that his url will not work.

File: lib/widgets/organization_list.dart (hunk 1 above)
Comment (SiddheshKukade) [reply lib/widgets/organization_list.dart]: Ok! got it. Go ahead.

**Comment by SiddheshKukade** [comment]
@Ayush0Chaudhary  Yes I got it thanks, changes are looking fine to me now.

**Review by SiddheshKukade — APPROVED** [review]

**Comment by Ayush0Chaudhary** [comment]
@palisadoes @SiddheshKukade  Please ask @Sagar2366  for 1 more review?
When I asked for re-review from @SiddheshKukade his review got removed?

**Comment by SiddheshKukade** [comment]
> @palisadoes @SiddheshKukade Please ask @Sagar2366 for 1 more review? When I asked for re-review from @SiddheshKukade his review got removed?

@Ayush0Chaudhary I have already reviewed your changes.

**Comment by Ayush0Chaudhary** [comment]
but it is showing At least 2 approving reviews are required by reviewers with write access for merge

**Comment by palisadoes** [comment]
@Ayush0Chaudhary The code testing coverage has been reduced by -2.51% with this change. Please add the necessary tests to make sure that the coverage is maintained.

**Comment by palisadoes** [comment]
@Ayush0Chaudhary Is there a standardized function/class for all error messages? It would greatly help with creating a standardized look and feel for the app. Then we could create an issue to migrate all errors to the new methodology.

Is there a good best practice that we could follow for this?

**Comment by SiddheshKukade** [comment]
> @Ayush0Chaudhary Is there a standardized function/class for all error messages? It would greatly help with creating a standardized look and feel for the app. Then we could create an issue to migrate all errors to the new methodology.
> 
> Is there a good best practice that we could follow for this?

@palisadoes Sir, we can create a separate custom widget for it and we can name that as `TalawaErrorDialogue` which will take a string to show the error message on the screen

**Comment by palisadoes** [comment]
> > @Ayush0Chaudhary Is there a standardized function/class for all error messages? It would greatly help with creating a standardized look and feel for the app. Then we could create an issue to migrate all errors to the new methodology.
> > Is there a good best practice that we could follow for this?
> 
> @palisadoes Sir, we can create a separate custom widget for it and we can name that as `TalawaErrorDialogue` which will take a string to show the error message on the screen

OK, @Ayush0Chaudhary please add this to this PR then we can open another issue to convert all our errors to use the new widget.

**Comment by Ayush0Chaudhary** [comment]
OK @palisadoes sir, I'll add this Widget

**Comment by Ayush0Chaudhary** [comment]
also we can specify custom error messages like below, just example 
```

class Constants {
  /// USER AUTHENTICATION CONSTANTS
  static const String USER_AUTH_WRONG_CREDENTIALS =
      'Invalid Credentials, Please check.';
  static const String USER_AUTH_USER_NOT_FOUND =
      'User Not found, try Signing Up.';
  static const String USER_NOT_AUTHORIZED_TO_FETCH_USER =
      'You are not authorized to fetch this User';
  static const String USER_NOT_FOUND = 'No User Found';

  /// GENERIC FAILURE CONSTANTS
  static const String BAD_RESPONSE_FORMAT = 'Bad Response Format';
  static const String GENERIC_FAILURE =
      'Something Wrong Occured! Please try again.';
  static const String NO_INTERNET_CONNECTION = 'No Internet Connection';
  static const String HTTP_EXCEPTION = 'Unable to fetch response';
  static const String UNAUTHORIZED_EXCEPTION =
      'Unauthorized to fetch the resource.';
}

```

@palisadoes @SiddheshKukade

**Comment by palisadoes** [comment]
> also we can specify custom error messages like below, just example
> 
> ```
> 
> class Constants {
>   /// USER AUTHENTICATION CONSTANTS
>   static const String USER_AUTH_WRONG_CREDENTIALS =
>       'Invalid Credentials, Please check.';
>   static const String USER_AUTH_USER_NOT_FOUND =
>       'User Not found, try Signing Up.';
>   static const String USER_NOT_AUTHORIZED_TO_FETCH_USER =
>       'You are not authorized to fetch this User';
>   static const String USER_NOT_FOUND = 'No User Found';
> 
>   /// GENERIC FAILURE CONSTANTS
>   static const String BAD_RESPONSE_FORMAT = 'Bad Response Format';
>   static const String GENERIC_FAILURE =
>       'Something Wrong Occured! Please try again.';
>   static const String NO_INTERNET_CONNECTION = 'No Internet Connection';
>   static const String HTTP_EXCEPTION = 'Unable to fetch response';
>   static const String UNAUTHORIZED_EXCEPTION =
>       'Unauthorized to fetch the resource.';
> }
> ```
> 
> @palisadoes @SiddheshKukade

Remember that we need to ensure that the errors work in the selected language that user selects. The widget and supporting code / data will to be adjusted to support this.

**Comment by Ayush0Chaudhary** [comment]
@SiddheshKukade their already exist a function to show snack bar
https://github.com/PalisadoesFoundation/talawa/blob/46cf4efd53cbdbdba3e42633875c667eff4a94cf/lib/services/navigation_service.dart#L52-L63

Can you explain what you meant in

**Comment by Ayush0Chaudhary** [comment]
errorw
@palisadoes @SiddheshKukade is this good??

**Comment by palisadoes** [comment]
1. Does it support multilingual functionality? If not you'll have to also update the API code with the translations.
2. Does this widget support multi-line and multi-paragraph messages that allows the user to scroll if necessary?

We really need to ensure that the app's user experience is improved. This is a very good first step and I encourage you to go the extra mile.

**Comment by Ayush0Chaudhary** [comment]
> 1. Does it support multilingual functionality? If not you'll have to also update the API code with the translations.
> 2. Does this widget support multi-line and multi-paragraph messages that allows the user to scroll if necessary?
> 
> We really need to ensure that the app's user experience is improved. This is a very good first step and I encourage you to go the extra mile. 

No for both answers, I just made a widget, but I'll implement it ASAP

**Comment by Ayush0Chaudhary** [comment]
https://user-images.githubusercontent.com/95746190/211491751-d36a9abd-3c6d-492a-bd22-c329cd636e60.mp4

Added the Allow scroll feature, working on multilingual messages

**Review by palisadoes — CHANGES_REQUESTED** [review]
@Ayush0Chaudhary The code coverage level for the repo has declined due to these changes. Please add tests that cover the work you have done.

image

**Comment by palisadoes** [comment]
@Ayush0Chaudhary have you been able to make any progress? Do you need guidance on tests from other contributors?

**Comment by Ayush0Chaudhary** [comment]
https://github.com/PalisadoesFoundation/talawa/pull/1382#issuecomment-1378193784 


 I wrote about the problem I had with tests, can you help me??
@palisadoes @SiddheshKukade

**Comment by palisadoes** [comment]
@SiddheshKukade Could you provide assistance?

**Comment by palisadoes** [comment]
@TheHazeEffect Please take a look at this too.

**Comment by TheHazeEffect** [comment]
> scrl.mp4 
> Added the Allow scroll feature, working on multilingual messages

**Review by palisadoes — CHANGES_REQUESTED** [review]
This is much better. Please write test cases for the code you have added. There are many examples in our code base. Please work with @SiddheshKukade and @TheHazeEffect (Troy Anderson) as mentors on this. Both are on slack.

File: lib/services/navigation_service.dart [hunk 2]
```diff
@@ -62,6 +64,78 @@ class NavigationService {
     );
   }
 
+  void showTalawaErrorWidget(
+    String message, {
+    Duration duration = const Duration(seconds: 2),
+  }) {
+    ScaffoldMessenger.of(navigatorKey.currentContext!).showSnackBar(SnackBar(
+      padding: EdgeInsets.zero,
```
Comment (TheHazeEffect) [inline lib/services/navigation_service.dart:72]: Could you refactor this into a reusable stateless widget that takes message as the parameter?
That way we can ensure snackbars are consistent across the app

File: lib/services/navigation_service.dart [hunk 3]
```diff
@@ -62,6 +64,78 @@ class NavigationService {
...
+          )
+        ],
+      ),
+      // backgroundColor: Colors.grey,
+      backgroundColor: const Color.fromRGBO(65, 65, 66, 1),
+    ));
+  }
+
+  void showTalawaErrorDialog(String errorMessage) {
+    showDialog(
```
Comment (TheHazeEffect) [inline lib/services/navigation_service.dart:109]: Do the same with this error dialog please.

**Review by TheHazeEffect — CHANGES_REQUESTED** [review]
A great PR.

Just add some of the changes I recommended and we should be a go

**Comment by TheHazeEffect** [comment]
> Screenshot from 2023-01-20 03-19-46
> 
> Screenshot from 2023-01-20 03-21-11
> 
> **I added the custom error widget** as @TheHazeEffect suggested.
> 
> @palisadoes @SiddheshKukade review please!!

Good job!

File: lib/services/navigation_service.dart (hunk 2 above)
Comment (TheHazeEffect) [reply lib/services/navigation_service.dart:72]: You also probably want to change to a auto sized text as well, incase the message is long or complex

Comment (Ayush0Chaudhary) [reply lib/services/navigation_service.dart:75]: I made the changes you asked for, please  check the PR

File: lib/services/navigation_service.dart (hunk 3 above)
Comment (Ayush0Chaudhary) [reply lib/services/navigation_service.dart:85]: Done the same for it too.

**Comment by noman2002** [comment]
@Ayush0Chaudhary `flutter analyze --no-pub` is failing. Please fix it by running `flutter format .` or `dart format .`

**Comment by noman2002** [comment]
@Ayush0Chaudhary pull the latest changes from develop.

**Comment by Ayush0Chaudhary** [comment]
@noman2002 I did as you asked.

**Comment by noman2002** [comment]
@Ayush0Chaudhary code coverage has gone down to 68.0 which is less than the minimum coverage required to merge a pr. Please try achieve 100% coverage for the file you have modified.

**Comment by Ayush0Chaudhary** [comment]
Screenshot from 2023-01-26 22-17-02

@noman2002 code coverage is 68.7 % now
I added the test for both the widget.

**Comment by Ayush0Chaudhary** [comment]
@palisadoes @noman2002 @TheHazeEffect @SiddheshKukade please review the PR

**Comment by noman2002** [comment]
@Ayush0Chaudhary As the codecov is not getting triggered please attach a screenshot of code coverage of the files you have made changes, to get this PR approved. It will be easier for mentors to review.

**Comment by Ayush0Chaudhary** [comment]
Screenshot from 2023-01-28 02-33-01
Screenshot from 2023-01-28 02-33-07

**Comment by Ayush0Chaudhary** [comment]
**For Talawa Error Widget (custom snackbar)**
Screenshot from 2023-01-28 02-35-56

**Comment by Ayush0Chaudhary** [comment]
**For talawa Error Dialog**
Screenshot from 2023-01-28 02-36-11

**Comment by Ayush0Chaudhary** [comment]
**For navigation service**
Screenshot from 2023-01-28 02-39-01

the part I added

**Comment by Ayush0Chaudhary** [comment]
Overall coverage increased to **68.7%** now
Screenshot from 2023-01-28 02-41-40

**Comment by Ayush0Chaudhary** [comment]
@noman2002 can we merge this PR now??

**Comment by Ayush0Chaudhary** [comment]
@noman2002 I added the code coverage of the test I added.
@palisadoes @TheHazeEffect @SiddheshKukade

**Comment by noman2002** [comment]
@Ayush0Chaudhary  Please remove the firebase api files it includes your api keys. Rest everything looks good to me.

**Comment by Ayush0Chaudhary** [comment]
@noman2002 done!! can we merge it now??

**Comment by noman2002** [comment]
> @noman2002 done!! can we merge it now??

Do we need to change anything in the installation process?

**Comment by Ayush0Chaudhary** [comment]
> > @noman2002 done!! can we merge it now??
> 
> Do we need to change anything in the installation process?

no, the process remains the same

**Review by noman2002 — APPROVED** [review]
@palisadoes Looks good to me. Please have a look at it as there are various files changed.

**Review by palisadoes — APPROVED** [review]
